File: leetcode/tree/113. Path Sum II/js/main.js
/**
 * @param node
 * @param {number} sum
 * @return {number[][]}
 */
const pathSum = function (node, sum) {
  let path = [];
  let paths = [];

  function findPaths(root, sum, path, paths) {
    if (!root) {
      return;
    }

    path.push(root.val);

    if (root.left === null && root.right === null && sum === root.val) {
      paths.push([...path]);
      path.pop();
      return;
    }

    findPaths(root.left, sum - root.val, path, paths);
    findPaths(root.right, sum - root.val, path, paths);

    path.pop();
  }

  findPaths(node, sum, path, paths);

  return paths;
};

File: leetcode/tree/113. Path Sum II/py/main.py
# Time O(N)
# Space O(N)
class Solution:
    def pathSum(self, root, target):
        if not root:
            return []
        
        def isLeaf(node):
            return not node.left and not node.right
        
        ans = []
  
        def dfs(node, target, comb):
          if node == None:
            return
          
          comb.append(node.val)
        
          if isLeaf(node) and target - node.val == 0:
            ans.append(list(comb))
            comb.pop()
            return

          dfs(node.left, target - node.val, comb)
          dfs(node.right, target - node.val, comb)
            
          comb.pop()

        dfs(root, target, [])
                
        return ans